libs/vkd3d/command.c

```diff
@@ -1412,35 +1421,117 @@ static HRESULT STDMETHODCALLTYPE d3d12_shared_fence_GetDevice(d3d12_fence_iface
 static UINT64 STDMETHODCALLTYPE d3d12_shared_fence_GetCompletedValue(d3d12_fence_iface *iface)
 {
     struct d3d12_shared_fence *fence = shared_impl_from_ID3D12Fence1(iface);
-    const struct vkd3d_vk_device_procs *vk_procs = &fence->device->vk_procs;
+
+    TRACE("iface %p\n", iface);
+    return vkd3d_atomic_uint64_load_explicit(&fence->last_value, vkd3d_memory_order_acquire);
+}
+
+static void vkd3d_shared_fence_signal_events(struct d3d12_shared_fence *fence, uint64_t value)
+{
+    struct vkd3d_shared_fence_waiting_event *waiting_event = NULL;
+    struct list *elem;
+
+    vkd3d_atomic_uint64_store_explicit(&fence->last_value, value, vkd3d_memory_order_release);
+
+    pthread_mutex_lock(&fence->mutex);
+    while (!list_empty(&fence->events))
+    {
+        elem = list_tail(&fence->events);
+        waiting_event = LIST_ENTRY(elem, struct vkd3d_shared_fence_waiting_event, entry);
+        if (waiting_event->value <= value)
+        {
+            fence->device->signal_event(waiting_event->event);
+            list_remove(elem);
+            vkd3d_free(waiting_event);
+        }
+        else
+        {
+            break;
+        }
+    }
+    pthread_mutex_unlock(&fence->mutex);
+}
+
+static void *vkd3d_shared_fence_worker_main(void *userdata)
+{
+    const struct vkd3d_vk_device_procs *vk_procs;
+    struct d3d12_shared_fence *fence;
+    VkSemaphoreWaitInfo wait_info;
     uint64_t completed_value;
     VkResult vr;
 
-    TRACE("iface %p\n", iface);
+    fence = userdata;
+    vk_procs = &fence->device->vk_procs;
+
+    wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO;
+    wait_info.pNext = NULL;
+    wait_info.flags = 0;
+    wait_info.semaphoreCount = 1;
+    wait_info.pSemaphores = &fence->timeline_semaphore;
+    wait_info.pValues = NULL;
 
-    vr = VK_CALL(vkGetSemaphoreCounterValueKHR(fence->device->vk_device, fence->timeline_semaphore, &completed_value));
-    if (vr != VK_SUCCESS)
+    while (!vkd3d_atomic_uint32_load_explicit(&fence->stop, vkd3d_memory_order_acquire))
     {
-        ERR("Failed to get shared fence counter value, error %d.\n", vr);
-        return 0;
+        vr = VK_CALL(vkGetSemaphoreCounterValueKHR(fence->device->vk_device, fence->timeline_semaphore, &completed_value));
+        if (vr != VK_SUCCESS)
+        {
+            ERR("Failed to get shared fence counter value, error %d.\n", vr);
+            return NULL;
+        }
+
+        vkd3d_shared_fence_signal_events(fence, completed_value);
+
+        if (!!vkd3d_atomic_uint32_load_explicit(&fence->stop, vkd3d_memory_order_acquire))
+            return NULL;
+
+        completed_value++;
```

Comment:
This is questionable in face of rewind. I would expect that you'd want to push wait requests to a queue. The wait thread wakes up, blocks on vkGetSemaphoreCounterValueKHR, signals the event and goes to sleep again.